AppCenter/AppCenterTests/MSByoiTests.m

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+#import "MSAppCenter.h"
+#import "MSAppCenterPrivate.h"
+#import "MSAuthTokenContext.h"
+#import "MSAuthTokenContextDelegateWrapper.h"
+#import "MSAuthTokenContextPrivate.h"
+#import "MSAuthTokenValidityInfo.h"
+#import "MSTestFrameworks.h"
+
+static NSString *const kMSJwtFormat = @"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.%@";
+
+@interface MSJwtHelper : NSObject
+
++ (NSString *)createJwtWithUserId:(NSString *)userId expiration:(int)expiration;
+
+@end
+
+@implementation MSJwtHelper
+
++ (NSString *)createJwtWithUserId:(NSString *)userId expiration:(int)expiration {
+  NSString *jsonClaims = [NSString stringWithFormat:@"{\"sub\":\"%@\",\"exp\":\"%i\"}", userId, expiration];
+  NSData *nsdata = [jsonClaims dataUsingEncoding:NSUTF8StringEncoding];
+  NSString *base64Encoded = [nsdata base64EncodedStringWithOptions:0];
+  NSString *combinedJwt = [NSString stringWithFormat:kMSJwtFormat, base64Encoded];
+  return combinedJwt;
+}
+
++ (NSString *)createJwtWithMissingExpClaim {
+  NSString *userId = @"some_user_id";
+  NSString *jsonClaims = [NSString stringWithFormat:@"{\"sub\":\"%@\"}", userId];
+  NSData *nsdata = [jsonClaims dataUsingEncoding:NSUTF8StringEncoding];
+  NSString *base64Encoded = [nsdata base64EncodedStringWithOptions:0];
+  return [NSString stringWithFormat:kMSJwtFormat, base64Encoded];
+}
+
+@end
+
+@interface MSByoiTests : XCTestCase
+
+@property(nonatomic) MSAppCenter *sut;
+@property(nonatomic) id authTokenContextMock;
+
+@end
+
+@implementation MSByoiTests
+
+- (void)setUp {
+  [super setUp];
+
+  // Auth token context.
+  [MSAuthTokenContext resetSharedInstance];
+  self.authTokenContextMock = OCMPartialMock([MSAuthTokenContext new]);
+  OCMStub([self.authTokenContextMock sharedInstance]).andReturn(self.authTokenContextMock);
+
+  // System Under Test.
+  [MSAppCenter resetSharedInstance];
+  self.sut = [[MSAppCenter alloc] init];
+}
+
+- (void)tearDown {
+  [self.authTokenContextMock stopMocking];
+  [MSAuthTokenContext resetSharedInstance];
+  [super tearDown];
+}
+
+- (void)testSetAuthTokenDelegateValidToken {
+
+  // If
+  // Valid Token
+  NSString *userId = @"original_user_id";
+  NSString *currentJwt = [MSJwtHelper createJwtWithUserId:userId expiration:86400];
+  MSAuthTokenValidityInfo *validityInfo =
+      [[MSAuthTokenValidityInfo alloc] initWithAuthToken:currentJwt
+                                               startTime:nil
+                                                 endTime:[[NSDate alloc] initWithTimeIntervalSinceNow:86400]];
+  [MSAppCenter setAuthToken:currentJwt];
+
+  // Stub delegate mock.
+  id delegateMock = OCMProtocolMock(@protocol(MSAuthTokenDelegate));
+  __block int numCalls = 0;
+  OCMStub([delegateMock appCenter:OCMOCK_ANY acquireAuthTokenWithCompletionHandler:OCMOCK_ANY]).andDo(^(__unused NSInvocation *invocation) {
```

Comment:
Use `OCMReject` for verification if you are verifying the method not to be called.